### ISSUE TITLE
no-mixed-spaces-and-tabs with smart tabs does not work as expected

### ISSUE THREAD

**Issue by MethodGrab** [eslint/eslint#7323]
**Tell us about your environment**
- **ESLint Version:** 3.7.1
- **Node Version:** 6.3.0
- **npm Version:** 3.10.5

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```
/* eslint-env es6 */
/* eslint indent: [ 'error', 'tab' ], no-mixed-spaces-and-tabs: [ 'error', 'smart-tabs' ] */
```

**What did you do? Please include the actual source code causing the issue.**

[Also available as a Gist](https://gist.github.com/MethodGrab/3670dba9f8461385bae0d62a2864631c) with preserved indentation & reproducible in the [online demo](http://eslint.org/demo).

``` js
/* eslint-env es6 */
/* eslint indent: [ 'error', 'tab' ], no-mixed-spaces-and-tabs: [ 'error', 'smart-tabs' ] */

const t1 = (  ) => {
    // should not warn
    // `→   return 't1';`
    return 't1';
};

const t2 = (  ) => {
    // should warn about mixed spaces and tabs
    // `→   ·return 't2';`
     return 't2';
};


// should warn about incorrect indent level
// `→   let t3;`
    let t3;

const b = (  ) => {

    // should warn about incorrect indent type (4 spaces instead of 1 tab)
    // `····let t4;`
    let t4;

    if ( true ) {
        // should warn about mixed spaces and tabs
        // `→   ····let t5;`
        let t5;
    }

}
```

**What did you expect to happen?**
Mixed spaces and tabs that are not used for alignment (eg`t2` & `t5` above) should trigger a `no-mixed-spaces-and-tabs` warning.

If the `smart-tabs` option is removed, it triggers `no-mixed-spaces-and-tabs` warnings as expected.

``` js
/* eslint-env es6 */
/* eslint indent: [ 'error', 'tab' ], no-mixed-spaces-and-tabs: [ 'error' ] */
```

**What actually happened? Please include the actual, raw output from ESLint.**

```
eslint v3.7.1
  19:2  error  Expected indentation of 0 tabs but found 1        indent
  25:5  error  Expected indentation of 1 tab but found 4 spaces  indent
```

eslint v3.6.1 actually caught this in the `indent` rule (possibly related to #7248).

```
eslint v3.6.1
  13:3  error  Expected indentation of 1 tab but found 1 space and 1 tab    indent
  19:2  error  Expected indentation of 0 tabs but found 1                   indent
  25:5  error  Expected indentation of 1 tab but found 4 spaces             indent
  30:6  error  Expected indentation of 2 tabs but found 4 spaces and 1 tab  indent
```

**Comment by not-an-aardvark** [comment]
Thanks for the report.

I'm a bit confused about the intended behavior here. You mentioned that the rule should report "Mixed spaces and tabs that are not used for alignment". Why is this the case for t2 and t5 but not any of the other examples?

**Comment by MethodGrab** [comment]
Sorry that test case probably isn't minimal enough, I was trying to work out which rule was causing the problem. t2 & t5 are the only examples with mixed spaces and tabs, t1 is valid and t3 & t4 are caught by the indent rule. I've updated the gist:

```js
/* eslint-env es6 */
/* eslint indent: [ 'error', 'tab' ], no-mixed-spaces-and-tabs: [ 'error', 'smart-tabs' ] */

const t2 = (  ) => {
	// should warn about mixed spaces and tabs
	// `→   ·return 't2';`
	 return 't2';
};

const b = (  ) => {

	if ( true ) {
		// should warn about mixed spaces and tabs
		// `→   ····let t5;`
	    let t5;
	}

}
```

**Comment by not-an-aardvark** [comment]
I see, thanks for clarifying.

I think in those cases, the rule is just verifying that the line has a block of tabs followed by a block of spaces, rather then some other mixture of spaces and tabs. Since you have `smart-tabs` enabled, it assumes that the spaces are being used for alignment, so it doesn't report any errors.

Personally, I think there are a lot of different points of view on how many spaces should be used for alignment in any given case, so the rule shouldn't mandate any particular amount. (Admittedly, I don't use smart tabs in my own code, so perhaps the issue is clearer than I think it is.)

**Comment by MethodGrab** [comment]
>"smart-tabs" allows mixed spaces and tabs when the latter are used for alignment.
– http://eslint.org/docs/rules/no-mixed-spaces-and-tabs

Oh I see. From the wording of the docs (and the smart-tabs example) I assumed it checked surrounding statements to verify that it was actually being used for alignment. If that's not the case then you can close this.

**Comment by not-an-aardvark** [comment]
I think the resolution is that this is working as intended, so I'm closing this issue.
